server/bleep/src/state.rs

```diff
@@ -47,6 +48,95 @@ pub struct StateSource {
     cookie_key: Option<PathBuf>,
 }
 
+pub struct PersistedState<T> {
+    path: PathBuf,
+    state: Arc<T>,
+}
+
+impl<T: Serialize + DeserializeOwned + Default + Send + Sync> PersistedState<T> {
+    fn load_or_default(name: &'static str, source: &StateSource) -> Result<Self> {
+        let path = source.directory().join(name).with_extension("json");
+        Ok(Self {
+            state: Arc::new(read_file_or_default(&path)?),
+            path,
+        })
+    }
+
+    fn load_or(name: &'static str, source: &StateSource, val: T) -> Self {
+        let path = source.directory().join(name).with_extension("json");
+        Self {
+            state: Arc::new(read_file(&path).unwrap_or(val)),
+            path,
+        }
+    }
+
+    pub fn store(&self) -> Result<()> {
+        Ok(pretty_write_file(&self.path, self.state.as_ref())?)
+    }
+}
+
+impl<T> Deref for PersistedState<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.state
+    }
+}
+
+impl<T> Clone for PersistedState<T> {
+    fn clone(&self) -> Self {
+        Self {
+            path: self.path.clone(),
+            state: self.state.clone(),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ApplicationSeed(String);
+
+impl From<Option<String>> for ApplicationSeed {
+    fn from(value: Option<String>) -> Self {
+        match value {
+            Some(val) => ApplicationSeed(val),
+            None => Self::default(),
+        }
+    }
+}
+
+impl ToString for ApplicationSeed {
+    fn to_string(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+impl Default for ApplicationSeed {
+    fn default() -> Self {
+        let seed: [u8; 32] = rand::random();
```

Comment:
Why use this rather than `uuid`?

Reply:
There's no good reason, I didn't think of `uuid` at all. I can change it if that seems more consistent.